### ISSUE TITLE
JFrog is not working with TF 0.13-0.15

### ISSUE THREAD

**Issue by vnatarov-mn** [jfrog/terraform-provider-artifactory#76]
I could not run Terraform code, got the next issue:

```
$ tfswitch
Reading required version from terraform file, constraint: = 0.13
Matched version: 0.13.0
Downloading https://releases.hashicorp.com/terraform/0.13.0/terraform_0.13.0_darwin_amd64.zip to terraform_0.13.0_darwin_amd64.zip
Downloading ...
35631250 bytes downloaded.
Switched terraform to version "0.13.0"

$ terraform init
Initializing modules...

Initializing the backend...

Initializing provider plugins...
- Finding jfrog/artifactory versions matching "2.2.9"...
- Finding latest version of hashicorp/artifactory...
- Installing jfrog/artifactory v2.2.9...
- Installed jfrog/artifactory v2.2.9 (signed by a HashiCorp partner, key ID 6B219DCCD7639232)

Partner and community providers are signed by their developers.
If you'd like to know more about provider signing, you can read about it here:
https://www.terraform.io/docs/plugins/signing.html

Error: Failed to install provider

Error while installing hashicorp/artifactory: provider registry
registry.terraform.io does not have a provider named
registry.terraform.io/hashicorp/artifactory
```

TF code:
```
# Required for Terraform 0.13 and up (https://www.terraform.io/upgrade-guides/0-13.html)
terraform {

  # required_version = "= 0.13"

  required_providers {
    
    artifactory = {
      source = "jfrog/artifactory"
      version = "2.2.9"
    }
  }
}

# Configure the Artifactory provider
provider "artifactory" {
  url      = "https://internal-adb96e117fff1436ea9aedbaf0ca997b-774872943.us-east-1.elb.amazonaws.com"
  
  username = "admin"
  password = "admin"
  # or
  # access_token = null
  # api_key      = null

}

module "jfrog_local_repository" {
  source = "git@github.com:SebastianUA/terraform.git//jfrog/modules/jfrog_repository?ref=dev"

  enable_local_repository       = true
  local_repository_key          = "local-repo-1"
  local_repository_package_type = "npm"
}

module "jfrog_remote_repository" {
  source = "git@github.com:SebastianUA/terraform.git//jfrog/modules/jfrog_repository?ref=dev"

  enable_remote_repository       = true
  remote_repository_key          = "remote-repo-1"
  remote_repository_package_type = "npm"
  remote_repository_url          = "https://registry.npmjs.org/"

  remote_repository_repo_layout_ref = "npm-default"
}

module "jfrog_virtual_repository" {
  source = "git@github.com:SebastianUA/terraform.git//jfrog/modules/jfrog_repository?ref=dev"

  enable_virtual_repository       = true
  virtual_repository_key          = "virtual-repo-1"
  virtual_repository_package_type = "maven"
  virtual_repository_repositories = []
}
```

**Comment by chb0github** [comment]
Thanks for letting us know - we will investigate, time permitting.

**Comment by vrumiantsev** [comment]
Hello @chb0github , any updates about this bug?

**Comment by chb0github** [comment]
None, unfortunately - We won't be able to have a look until after the end-of-the-month.

**Comment by gkman** [comment]
There is no issue with the provider. 

Reading over this info about [provider inheritance](https://www.terraform.io/docs/language/modules/develop/providers.html#implicit-provider-inheritance) there is a note that states:

```txt
Only provider configurations are inherited by child modules, not provider source or version requirements. Each module must declare its own provider requirements. This is especially important for non-HashiCorp providers.
```

So you must define your `required_providers` in each defined module you create for anything that is a `non-HashiCorp` provider.

**Comment by chb0github** [comment]
I'm familiar with this error - what versions have you tried?

**Comment by chb0github** [comment]
No, I mean, what version(s) of the provider did you try?

**Comment by chb0github** [comment]
That's an issue. Thanks for the details

**Comment by gkman** [comment]
@vnatarov-mn Sorry for the delay, in this example you shared, I see you define the provider. And are trying to pass in the configuration. However where you define your [modules](https://github.com/SebastianUA/terraform/blob/master/jfrog/modules/jfrog_repository/versions.tf) you must have defined the required provider as I mentioned in my original comment.

So in `github.com/SebastianUA/terraform/blob/master/jfrog/modules/versions.tf` you need to have something like:

```hcl
terraform {
  required_version = "= 0.13.5"
  required_providers {
    artifactory = {
      source  = "jfrog/artifactory"
      version = "2.2.12"
    }
  }
}
```

And then in your usage you can have the following, and the configuration will be passed implicitly along with your module invocation.

```hcl
module "docker_draft_local" {
  source = "git@github.com:SebastianUA/terraform.git//jfrog/modules/jfrog_repository"

  ...
}

provider "artifactory" {
  url      = var.url
  username = var.username
  password = var.password
}
```

I have tested this myself on `terraform v0.13.5` and `jfrog/artifactory v2.2.12`

**Comment by vnatarov-mn** [comment]
@gkman, Thank you! As a workaround - it's fine, but that's the wrong way to write the code. I have found a new bug/feature that doesn't exist inside this Terraform provider.

**Comment by gkman** [comment]
I'm not sure I follow, this is what `HashiCorp Terraform` documentation expects and they explain why and I explained that [above](https://github.com/jfrog/terraform-provider-artifactory/issues/76#issuecomment-837418054)

**Comment by vnatarov-mn** [comment]
@gkman,  If I'm hardcoding some specific version of the software inside Terraform module with resources as you wrote, I will not be able to overwrite when defining the provider and module(s). Other guys are able to write clear code on GO to override the needed version of the software when you're using the TF module. Let me provide some additional examples:
- Heroku: https://github.com/SebastianUA/terraform/blob/master/heroku/examples/heroku_account/main.tf
- K8S: https://github.com/SebastianUA/terraform/blob/master/kubernetes/examples/k8s_cluster_role/main.tf
- Cloudflare: https://github.com/SebastianUA/terraform/blob/master/cloudflare/examples/cloudflare_record/versions.tf
- Others.

Of course, I will use your way to contribute everything that needs for Jfrog, but it's too hard to support the TF provider. 
 As for me, everyone's will fork that repo and fix issues with features. 
 
 If you will have any questions, feel free to contact me.
 --Vitalii

**Comment by gkman** [comment]
The only thing I've hard coded in my example is the required version of the provider my module needs to use. This is expected and required behavior from Terraform to handle dependency resolution and graph prior to running a `plan`. Similarly like I said I can pass one or even many providers to a module to configure it on the fly and how said provider operates.

So if I want to I can configure 2 different providers one for a `production` and one for a `staging` Artifactory with 2 different urls, usernames, and passwords. Then I can invoke the module and pass each invocation a different provider configuration. The only limitation is, again, I have to invoke it using the required provider version as defined by the module. 

I would definitely take the time to thoroughly read through [Implicit Provider Inheritance](https://www.terraform.io/docs/language/modules/develop/providers.html#implicit-provider-inheritance) and [Passing Providers Explicitly](https://www.terraform.io/docs/language/modules/develop/providers.html#passing-providers-explicitly)

What are you looking to do that is different than that pattern, and why?

**Comment by chb0github** [comment]
I can't verify any of this - there are references to internal AWS albs. If someone has a C&P example, and I can verify it, then I an fix it. If Not, I will just end up closing this ticket as not reproducible

**Comment by SebastianUA** [comment]
Hi @chb0github,

Yes, I can reproduce that issue:
```
# Required for Terraform 0.13 and up (https://www.terraform.io/upgrade-guides/0-13.html)
terraform {

  # required_version = "= 0.13"

  required_providers {
    
    artifactory = {
      source = "jfrog/artifactory"
      version = "2.2.9"
    }
  }
}

# Configure the Artifactory provider
provider "artifactory" {
  url      = "https://internal-adb96e117fff1436ea9aedbaf0ca997b-774872943.us-east-1.elb.amazonaws.com"
  
  username = "admin"
  password = "admin"
  # or
  # access_token = null
  # api_key      = null

}

module "jfrog_local_repository" {
  source = "git@github.com:SebastianUA/terraform.git//jfrog/modules/jfrog_repository?ref=jfrog_tmp"

  enable_local_repository       = true
  local_repository_key          = "local-repo-1"
  local_repository_package_type = "npm"
}

module "jfrog_remote_repository" {
  source = "git@github.com:SebastianUA/terraform.git//jfrog/modules/jfrog_repository?ref=jfrog_tmp"

  enable_remote_repository       = true
  remote_repository_key          = "remote-repo-1"
  remote_repository_package_type = "npm"
  remote_repository_url          = "https://registry.npmjs.org/"

  remote_repository_repo_layout_ref = "npm-default"
}

module "jfrog_virtual_repository" {
  source = "git@github.com:SebastianUA/terraform.git//jfrog/modules/jfrog_repository?ref=jfrog_tmp"

  enable_virtual_repository       = true
  virtual_repository_key          = "virtual-repo-1"
  virtual_repository_package_type = "maven"
  virtual_repository_repositories = []
}
``` 

Ran `terraform init` and got:
```
terraform init
Initializing modules...

Initializing the backend...

Initializing provider plugins...
- Finding jfrog/artifactory versions matching "2.2.9"...
- Finding latest version of hashicorp/artifactory...
- Installing jfrog/artifactory v2.2.9...
- Installed jfrog/artifactory v2.2.9 (signed by a HashiCorp partner, key ID 6B219DCCD7639232)

Partner and community providers are signed by their developers.
If you'd like to know more about provider signing, you can read about it here:
https://www.terraform.io/docs/cli/plugins/signing.html
╷
│ Error: Failed to query available provider packages
│
│ Could not retrieve the list of available versions for provider hashicorp/artifactory: provider registry registry.terraform.io does not have a provider named registry.terraform.io/hashicorp/artifactory
│
│ Did you intend to use jfrog/artifactory? If so, you must specify that source address in each module which requires that provider. To see which modules are currently depending on hashicorp/artifactory,
│ run the following command:
│     terraform providers
```

**Comment by chb0github** [comment]
Ok. I am able to reproduce. Let me dig into this.

**Comment by chb0github** [comment]
Right... I remember now. We don't support modules. Some members of my team reminded me of that. We can convert this to a feature

**Comment by SebastianUA** [comment]
@chb0github LOL, It's related to the bug inside Jfrog provider neither modules!
Okay, thanks anyway!

Cheers

**Comment by chb0github** [comment]
@SebastianUA - I welcome your insight as to where and what the bug might be. I spent the better part of a day trying to sort out what the problem might be

**Comment by chb0github** [comment]
@jasonwbarnett - comments?

**Comment by jasonwbarnett** [comment]
@chb0github 

This  issue is being caused by a version conflict between the top level provider version constraint declared in main.tf and the version pinning defined here: https://github.com/SebastianUA/terraform/blob/dev/jfrog/modules/jfrog_repository/versions.tf

See (copy and pasted from original issue description):

```
terraform {
  required_providers {
    
    artifactory = {
      source = "jfrog/artifactory"
      version = "2.2.9"
    }
  }
}
```

This is pinning hard to 2.2.9 and the [jfrog_repository module][1] has it pinned to ~>2.2.12 (>= 2.2.12 && < 2.3) defined.

```terraform
# pasted from jfrog_repositry module for brevity
terraform {
  required_providers {
    artifactory = {
      source  = "jfrog/artifactory"
      version = "~> 2.2.11"
    }
  }
  required_version = "~> 0.13"
}
```

In summary, the version constraints are at odds with each other.

- = 2.2.9
- ~> 2.2.11

[1]: https://github.com/SebastianUA/terraform/blob/dev/jfrog/modules/jfrog_repository/versions.tf

**Comment by chb0github** [comment]
@jasonwbarnett - that didn't even occur to me. Nice catch.

**Comment by jasonwbarnett** [comment]
I validated that changing the pinning (see below) it works just fine.

```terraform
terraform {
  required_providers {
    
    artifactory = {
      source = "jfrog/artifactory"
      version = "~> 2.2.9"
    }
  }
}
```

```bash
$ terraform init -upgrade
Upgrading modules...
Downloading git@github.com:SebastianUA/terraform.git?ref=dev for jfrog_local_repository...
- jfrog_local_repository in .terraform/modules/jfrog_local_repository/jfrog/modules/jfrog_repository
Downloading git@github.com:SebastianUA/terraform.git?ref=dev for jfrog_remote_repository...
- jfrog_remote_repository in .terraform/modules/jfrog_remote_repository/jfrog/modules/jfrog_repository
Downloading git@github.com:SebastianUA/terraform.git?ref=dev for jfrog_virtual_repository...
- jfrog_virtual_repository in .terraform/modules/jfrog_virtual_repository/jfrog/modules/jfrog_repository

Initializing the backend...

Initializing provider plugins...
- Finding jfrog/artifactory versions matching "~> 2.2.9, ~> 2.2.11"...
- Installing jfrog/artifactory v2.2.15...
- Installed jfrog/artifactory v2.2.15 (signed by a HashiCorp partner, key ID 6B219DCCD7639232)

Partner and community providers are signed by their developers.
If you'd like to know more about provider signing, you can read about it here:
https://www.terraform.io/docs/cli/plugins/signing.html

Terraform has made some changes to the provider dependency selections recorded
in the .terraform.lock.hcl file. Review those changes and commit them to your
version control system if they represent changes you intended to make.

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
```

**Comment by SebastianUA** [comment]
Hi @jasonwbarnett,

Of course, it will be worked :)
I hardcoded the code inside the module:
```
terraform {
  required_providers {
    
    artifactory = {
      source = "jfrog/artifactory"
      version = "~> 2.2.9"
    }
  }
}
```

Please try to use:
```
# Required for Terraform 0.13 and up (https://www.terraform.io/upgrade-guides/0-13.html)
terraform {

  # required_version = "= 0.13"

  required_providers {
    
    artifactory = {
      source = "jfrog/artifactory"
      version = "2.2.9"
    }
  }
}

# Configure the Artifactory provider
provider "artifactory" {
  url      = "https://internal-adb96e117fff1436ea9aedbaf0ca997b-774872943.us-east-1.elb.amazonaws.com"
  
  username = "admin"
  password = "admin"
  # or
  # access_token = null
  # api_key      = null

}

module "jfrog_local_repository" {
  source = "git@github.com:SebastianUA/terraform.git//jfrog/modules/jfrog_repository?ref=jfrog_tmp"

  enable_local_repository       = true
  local_repository_key          = "local-repo-1"
  local_repository_package_type = "npm"
}

module "jfrog_remote_repository" {
  source = "git@github.com:SebastianUA/terraform.git//jfrog/modules/jfrog_repository?ref=jfrog_tmp"

  enable_remote_repository       = true
  remote_repository_key          = "remote-repo-1"
  remote_repository_package_type = "npm"
  remote_repository_url          = "https://registry.npmjs.org/"

  remote_repository_repo_layout_ref = "npm-default"
}

module "jfrog_virtual_repository" {
  source = "git@github.com:SebastianUA/terraform.git//jfrog/modules/jfrog_repository?ref=jfrog_tmp"

  enable_virtual_repository       = true
  virtual_repository_key          = "virtual-repo-1"
  virtual_repository_package_type = "maven"
  virtual_repository_repositories = []
}
``` 
Thanks.

**Comment by jasonwbarnett** [comment]
@SebastianUA you can't define the same dependency in two different places with conflicting version constraints. The only way you may get this to work is if you specify an alias for the provider. See the docs here: https://www.terraform.io/docs/language/providers/configuration.html#alias-multiple-provider-configurations

**Comment by SebastianUA** [comment]
@jasonwbarnett, Not really understand why I should hardcode the next code inside my TF modules of the Jfrog provider (version.tf file inside the modules):
```
terraform {
  required_providers {
    
    artifactory = {
      source = "jfrog/artifactory"
      version = "~> 2.2.9"
    }
  }
}
``` 

When you guys will update the version of the software, I must update the TF modules to be updated with new features of the provider. Why? 

Also, When I'm using Terraform, I want to hardcode a TF version to be sure that the code is working properly with it and will use it on PROD.

I didn't have any issues with any provides before I started using Jfrog.

**Comment by jasonwbarnett** [comment]
@SebastianUA If you update your configuration to what you just replied with you will not need to update it each time a new version of the provider is released.

> ```terraform
> terraform {
>   required_providers {
>     
>     artifactory = {
>       source = "jfrog/artifactory"
>       version = "~> 2.2.9"
>     }
>   }
> }
> ```


You have a few options, but I'll give you my recommendation based on you indicating that you would like to maintain stability and not worry about changes in this provider breaking your code.

1. Update all version constraints to be the same thing. In the specific case you would pick **one** and only one of the following options and put it in all `required_providers` configurations for the artifactory provider.
   1. `= 2.2.9`
   2. `~> 2.2.11`
1. Loosen the pinning from equals (`=`) to pessimistic pinning (`~>`)

   > ```terraform
   > terraform {
   >   required_providers {
   >     
   >     artifactory = {
   >       source = "jfrog/artifactory"
   >       version = "~> 2.2.9"
   >     }
   >   }
   > }
   > ```


I'm going to assume you know what `~>` means, but if you don't here is some light reading to get you up to speed: https://thoughtbot.com/blog/rubys-pessimistic-operator
